version: enable Sorbet type checking

SeekingMeaning opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?
- [ ] Have you successfully run `brew man` locally and committed any changes?

-----

Notable changes:

- Added a `NullVersion` singleton class that extends `Version`
- Changed `Version::Token` to be an abstract class
- Added `version.rbi` to work around Sorbet limitation with aliases

There are a lot of `T.untyped` but this is a start, at least

reitermarkus commented on Library/Homebrew/version.rb:
```diff
@@ -330,11 +361,11 @@ def self._parse(spec, detected_from_url:)
       Pathname.new(spec.dirname).stem
     elsif spec_s.match?(/\.[^a-zA-Z]+$/)
-      Pathname.new(spec_s).basename
+      Pathname.new(spec_s).basename.to_s
```
```suggestion
      spec.basename.to_s
```

reitermarkus commented on Library/Homebrew/version.rb:
```diff
@@ -330,11 +361,11 @@ def self._parse(spec, detected_from_url:)
-      spec.basename
+      spec.basename.to_s
     elsif spec_s.match?(%r{((?:sourceforge\.net|sf\.net)/.*)/download$})
       Pathname.new(spec.dirname).stem
```
```suggestion
      spec.dirname.stem
```

reitermarkus commented on Library/Homebrew/version.rb:
```diff
@@ -317,11 +346,13 @@ def self.create(val)
     version.nil? ? NULL : new(version, detected_from_url: detected_from_url)
   end
 
+  sig { params(spec: T.untyped, detected_from_url: T::Boolean).returns(T.nilable(String)) }
```
```suggestion
  sig { params(spec: T.any(String, Pathname), detected_from_url: T::Boolean).returns(T.nilable(String)) }
```

reitermarkus commented on Library/Homebrew/version.rb:
```diff
@@ -317,11 +346,13 @@ def self.create(val)
     end
   end
 
+  sig { params(spec: T.untyped, detected_from_url: T::Boolean).returns(Version) }
```
```suggestion
  sig { params(spec: T.any(String, Pathname), detected_from_url: T::Boolean).returns(Version) }
```

reitermarkus commented on Library/Homebrew/version.rb:
```diff
@@ -160,7 +180,7 @@ def <=>(other)
         1
       when NullToken
-        -Integer(other <=> self)
+        -T.must(other <=> self)
```
Shouldn't this be `nil`, since it's not comparable?

SeekingMeaning replied:
See reply below (in Conversation tab) / above (in Files changed tab)

reitermarkus commented on Library/Homebrew/version.rb:
```diff
@@ -122,20 +138,22 @@ def inspect
         value <=> other.value
       when NumericToken, NullToken
-        -Integer(other <=> self)
+        -T.must(other <=> self)
```
Shouldn't this be `nil`, since it's not comparable?

SeekingMeaning replied:
No; `other <=> self` means `other.<=>(self)` which calls `NumericToken#<=>` (or `NullToken#<=>`)

reitermarkus replied:
Then I guess my question is: Shouldn't `NullToken#<=>` always return `nil`?

SeekingMeaning replied:
No, since `NullToken` is used for e.g. a comparison between `1.5` and `1.5.1` where `Version.new("1.5").patch` returns `nil`, which becomes `NULL_TOKEN` in `<=>` after calling `Token.from`

reitermarkus replied:
Ah, I see.

reitermarkus commented on Library/Homebrew/version.rb:
```diff
@@ -36,6 +40,7 @@ def self.create(val)
       end.new(val)
     end
 
+    sig { params(val: T.untyped).returns(T.nilable(Token)) }
```
The `else` contains an `if`, so it can be `nil`.

MikeMcQuaid replied:
Ah, gotcha, misread that, thanks.

MikeMcQuaid commented on Library/Homebrew/os/mac/version.rb:
```diff
@@ -76,7 +76,7 @@ def <=>(other)
         @to_sym ||= begin
           # Big Sur is 11.x but Catalina is 10.15.
-          major_macos = if major >= 11
+          major_macos = if T.must(major) >= 11
```
Sorbet question: what does this do and why is it needed?

reitermarkus replied:
It casts `T.nilable(Type)` to `Type`, ensuring something is non-`nil`.

MikeMcQuaid replied:
What happens if it is `nil` with or without Sorbet enabled at runtime?

reitermarkus replied:
Without, the `>=` call will fail; with, a `TypeError` is raised.

Both should never happen since the macOS major version should always be non-`nil`.

MikeMcQuaid replied:
Thanks. Is it possible to somehow move logic like this into where `major` is set/returned instead?

reitermarkus replied:
It's calling `Version#major`, which definitely can have a `nil` major version.

SeekingMeaning replied:
One option is to add an RBI file that says `OS::Mac::Version#major` will always return a non-nil `Token`

MikeMcQuaid replied:
> One option is to add an RBI file that says `OS::Mac::Version#major` will always return a non-nil `Token`

Something like that seems good 👍🏻

SeekingMeaning commented on Library/Homebrew/version.rb:
```diff
@@ -11,16 +11,20 @@ class Version
 
     include Comparable
 
+    sig { params(val: T.untyped).returns(Token) }
```
No, since non-`String` objects could still implement `#to_str`

reitermarkus replied:
The current implementation suggests that, but do we actually call it with anything other than `String`? I'd say we make this take a `String` and adjust the call sites if needed or add other supported types explicitly. Having a bunch of methods that take `T.untyped` kinda defeats the purpose of type-checking.

SeekingMeaning commented on Library/Homebrew/version.rb:
```diff
@@ -46,29 +51,38 @@ def self.from(val)
       end
     end
 
+    sig { returns(T.untyped) }
     attr_reader :value
 
+    sig { params(value: T.untyped).void }
     def initialize(value)
       @value = value
     end
 
+    sig { abstract.params(other: T.untyped).returns(T.nilable(Integer)) }
+    def <=>(other); end
```
`Token.from(other)` converts `T.untyped` to `T.nilable(Token)`

`<=>` returns `nil` if the converted token is `nil`

SeekingMeaning commented:
- Replaced `T.untyped` with `String` for `Token#create`, `Version#create`, and `Version#initialize`
  - Added `T.must` to `ENV["HOMEBREW_LINUX_MINIMUM_GLIBC_VERSION"]`
- Replaced `when nil` and `else` branches in `Token#from` with `return NULL_TOKEN if ...` guard clause
- Added `#null?` to `Token` class as overridable method
  - The `NullToken` class already had `#null?` but now we can add `override` to its `sig`

reitermarkus commented on Library/Homebrew/os/linux/glibc.rb:
```diff
@@ -22,7 +22,7 @@ def system_version
       sig { returns(Version) }
       def minimum_version
-        Version.new ENV["HOMEBREW_LINUX_MINIMUM_GLIBC_VERSION"]
+        Version.new(T.must(ENV["HOMEBREW_LINUX_MINIMUM_GLIBC_VERSION"]))
```
```suggestion
        Version.new(ENV.fetch("HOMEBREW_LINUX_MINIMUM_GLIBC_VERSION"))
```

MikeMcQuaid commented on Library/Homebrew/version.rb:
```diff
@@ -36,39 +40,49 @@ def self.create(val)
+    sig { returns(T.untyped) }
     attr_reader :value
 
+    sig { params(value: T.untyped).void }
```
Could we restrict this perhaps to `Token`, `String`, `Integer`, `nil` or `NullToken`? Would ease some other code, too.

SeekingMeaning replied:
I gave this a go, but I came to the conclusion that adding a type to this would require something like Java's generics where a subclass can specify the type of a parent variable. I don't think Sorbet supports this, though

SeekingMeaning replied:
1. Added a type for the `value` parameter for `#initialize`
    - `Token` &rarr; `T.nilable(T.any(String, Integer))`
    - `NullToken` has no parameters for `#initialize`
    - `StringToken` &rarr; `String`
    - `NumericToken` &rarr; `T.any(String, Integer)` (converted to `Integer` by `value.to_i`)
1. Added overrides for `attr_reader :value` with the following return types:
    - `NullToken` &rarr; `NilClass` (The class for plain `nil` is `NilClass`)
    - `StringToken` &rarr; `String`
    - `NumericToken` &rarr; `Integer`
1. Replaced `@value = value` with `@value = T.let(value, T.untyped)`, otherwise we get this error for `NullToken` (and a similar error for `StringToken` and `NumericToken`):

```
version.rb:107: Expected NilClass but found T.nilable(T.any(String, Integer)) for method result type https://srb.help/7005
     107 |    attr_reader :value
                           ^^^^^
  Expected NilClass
    version.rb:107: Method value has return type NilClass
     107 |    attr_reader :value
              ^^^^^^^^^^^^^^^^^^
  Got T.nilable(T.any(String, Integer)) originating from:
    version.rb:59:
    59 |      @value = value
              ^^^^^^
```

MikeMcQuaid commented on Library/Homebrew/version.rb:
```diff
@@ -145,12 +168,14 @@ class NumericToken < Token
     PATTERN = /[0-9]+/i.freeze
     extend T::Sig
 
+    sig { params(value: T.untyped).void }
```
Same here (using a specific list of types)
